### DISCUSSION TITLE
document session.execute better

### DISCUSSION THREAD

**Discussion by bluebrown** [sqlalchemy/sqlalchemy#7405] [Ideas]
Hi, I am not able to find out from the documentation how I am supposed to use the result of session.execute. 

I found this https://docs.sqlalchemy.org/en/14/orm/session_basics.html#querying-2-0-style, but it's not useful to me. I don't understand how I get usable data from this API. 

Furthermore, I feel like its a bit too hard to "just get data" with it. What am I supposed to do? Currently, I have a list of model references. How do I get data from it as list of dicts  and list of lists?

```
with Session(engine) as s:
    q = select(WarrantyRequest)
    res = s.execute(q)
    keys = res.keys()
    data_list = res.all()
    print(keys)      # should print list of column names
    print(data_list) # should print list of lists with row data
    dict_list = s.execute(q).mappings().all()
    print(dict_list) # should print list of dicts with column names as keys
```

The result is the below, which in its current form is of little use. I cannot see something in the documentation that describes how to make data from it.

```
RMKeyView(['WarrantyRequest'])
[(<models.mock.WarrantyRequest object at 0x7f4d065d3df0>,), ...]
[{'WarrantyRequest': <models.mock.WarrantyRequest object at 0x7f002b464df0>}, ... ]
```

**Comment by zzzeek** [comment]
it looks like you might need background on selecting ORM entities vs. specific ORM attributes.  See the section https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-orm-entities-and-attributes and https://docs.sqlalchemy.org/en/14/orm/queryguide.html#selecting-individual-attributes which discuss this general idea.   also in the tutorial at https://docs.sqlalchemy.org/en/14/tutorial/data_select.html#selecting-orm-entities-and-columns .

**Reply by bluebrown** [reply]
Hi, thanks for the reply. I have seen some of it already, but It's still hard to understand IMO. I am browsing the docs for quite some time now. Furthermore, I don't understand what's the point of getting a list of entities as a result. What cool things can we do with that?  Using this scalars' method also didn't change anything for me.

**Reply by zzzeek** [reply]
typically the whole reason one uses an ORM is so that a user-defined entities can be stored and retrieved from a database in their complete form, in a similar way as one might use an object database.

if you don't need that functionality and you want to just work with rows and columns in a SQL-oriented fashion, then I'd skip using the ORM entirely, just use Core.      It's much more directly like a relational database as far as results are concerned.

**Reply by bluebrown** [reply]
Hm, OK, thank you for the advice. I will consider my options, but your suggestion sounds reasonable to me. Maybe at some point I will understand or see the need to use ORM and those entities.  

I think I just didn't realize that there is such a big difference between connection and session. I though using session is the clean way to do it, and using a connection is more low level and may have some drawbacks when it comes to connection pooling etc.

**Reply by zzzeek** [reply]
not really, the engine/ connection are using the same connection pool.   SQLAlchemy Core exists as a separate thing from the ORM for, well, basically people who don't like or "get" ORMs , after all I was like that for all the years before I wrote SQLAlchemy so I know the type :)  .   But also it's a useful separation of concerns even if one is using the ORM.
